fuzz/fuzz_types.ml

```diff
@@ -0,0 +1,440 @@
+  | VInt64 of int64
+  | VFloat of float
+  | VString of string
+  | VBytes of bytes
+  | VList of any_value list
+  | VArray of any_value array
+  | VOption of any_value option
+  | VPair of any_value * any_value
+  | VTriple of any_value * any_value * any_value
+  | VResult of (any_value, any_value) result
+  | VRecord of dyn_record
+  | VVariant of dyn_variant
+  | VEnum of dyn_variant
+
+(* Since we can't use "real" records and variants to test how Irmin serializes
+   and deserializes them, we instead represent records with a hashtable of all
+   their fields, and variants as the name of the current constructor and its value. *)
+and dyn_record = (string * (string, any_value) Hashtbl.t[@opaque])
+
+and dyn_variant = string * string * any_value
+
+(** Pretty-print an [any_value] value. This should be generated by ppx_deriving,
+    but we can't easily use it in combination with the custom rewriter. *)
+let rec show_any_value = function
+  | VUnit _ -> "VUnit ()"
+  | VBool b -> fmt "VBool %B" b
+  | VChar c -> fmt "VChar %C" c
+  | VInt i -> fmt "VInt %d" i
+  | VInt32 i -> fmt "VInt32 %ldl" i
+  | VInt64 i -> fmt "VInt64 %LdL" i
+  | VFloat f -> fmt "VFloat %F" f
+  | VString s -> fmt "VString %S" s
+  | VBytes b -> fmt "VBytes %S" (Bytes.to_string b)
+  | VList l ->
+      l |> List.map show_any_value |> String.concat "; " |> fmt "VList [%s]"
+  | VArray a ->
+      a
+      |> Array.to_list
+      |> List.map show_any_value
+      |> String.concat "; "
+      |> fmt "VArray [|%s|]"
+  | VOption (Some v) -> fmt "VOption (Some (%s))" (show_any_value v)
+  | VOption None -> "VOption None"
+  | VPair (a, b) -> fmt "VPair (%s, %s)" (show_any_value a) (show_any_value b)
+  | VTriple (a, b, c) ->
+      fmt "VTriple (%s, %s, %s)" (show_any_value a) (show_any_value b)
+        (show_any_value c)
+  | VResult (Ok v) -> fmt "VResult (Ok (%s))" (show_any_value v)
+  | VResult (Error v) -> fmt "VResult (Error (%s))" (show_any_value v)
+  | VRecord r -> fmt "VRecord (%s)" (show_dyn_record r)
+  | VVariant v -> fmt "VVariant (%s)" (show_dyn_variant v)
+  | VEnum v -> fmt "VEnum (%s)" (show_dyn_variant v)
+
+and show_dyn_record (n, t) =
+  fmt "{ name: %S; %s }" n
+    (String.concat "; "
+       (Hashtbl.fold
+          (fun k x xs -> fmt "%S: (%s)" k (show_any_value x) :: xs)
+          t []))
+
+and show_dyn_variant (n, c, v) =
+  fmt "{ name: %S; current_case: %S; value: %S }" n c (show_any_value v)
+
+(** Internal representation of Irmin types. *)
+type 'a t =
+  | TUnit : unit t
+  | TBool : bool t
+  | TChar : char t
+  | TInt : int t
+  | TInt32 : int32 t
+  | TInt64 : int64 t
+  | TFloat : float t
+  | TString : string t
+  | TBytes : bytes t
+  | TList : 'a t -> 'a list t
+  | TArray : 'a t -> 'a array t
+  | TOption : 'a t -> 'a option t
+  | TPair : 'a t * 'b t -> ('a * 'b) t
+  | TTriple : 'a t * 'b t * 'c t -> ('a * 'b * 'c) t
+  | TResult : 'a t * 'e t -> ('a, 'e) result t
+  (* Each of these lists must have 1 to 5 elements. *)
```

Comment:
Is there a reason for lists between 1 and 5 elements?

Reply:
@icristescu  I forgot to include the explanation from pascutto/irmin/pull/2/, sorry.

Essentially, we use a code rewriter (fuzz/rewriter/rewriter.ml) to replace the [%impl_record 5] and [%impl_variant 5] extension points with code to check dynamically-generated records (resp. variants) with up to 5 fields (resp. cases). This number can be adjusted at will, but the number of lines generated by impl_variant is exponential in the number of cases, so 5 seemed like a reasonable compromise.

As for the reason why we need to use a code rewriter in the first place, the issue is essentially that the public API for records and variants in `Irmin.Type` expects us to build functions whose signature (and, in particular, whose number of arguments) depends on the fields of the record or enum that we are building.

Here is an example from the documentation:
```ocaml
type menu = { restaurant : string; items : (string * int32) list }

let t =
  record "t" (fun restaurant items -> { restaurant; items })
  |+ field "restaurant" string (fun t -> t.restaurant)
  |+ field "items" (list (pair string int32)) (fun t -> t.items)
  |> sealr
```

As you can see, we need to pass a "constructor function" to `record` (here `fun restaurant items -> { restaurant; items }`) whose signature depends on the number and type of fields in the record. But, since the fuzzer is building record types dynamically, and it is not possible–to the best of my knowledge–to build functions whose signature depends on runtime information, we have a bit of a problem.

My compromise was to use code generation as a way to define "constructor functions" for all possible signatures (or, to be exact, for all possible number of arguments, since the type of each argument is made generic by hiding it inside a GADT) at build time. This does mean, however, that we can't test dynamic records, variants or enums whose number of fields exceeds the number of generated "constructor functions", hence the limit of 5 here.